fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java

```diff
@@ -1393,18 +1405,61 @@ private OptExpression tryUnionRewrite(RewriteContext rewriteContext,
         return createUnion(queryInput, viewInput, rewriteContext);
     }
 
+    // retry to rewrite the rewritten failed predicates by enforcing the columns not exists in query
+    private Pair<ScalarOperator, OptExpression> tryRewritePredicates(
+            RewriteContext rewriteContext,
+            ColumnRefSet mvToQueryRefSet,
+            OptExpression query,
+            ColumnRefFactory queryRefFactory,
+            List<ScalarOperator> predicatesToEnforce) {
+        List<ColumnRefOperator> columns = Lists.newArrayList();
+        for (ScalarOperator predicate : predicatesToEnforce) {
+            predicate.getColumnRefs(columns);
+        }
+        // check every columns to enforce should be in the Scan Operator
+        for (ColumnRefOperator column : columns) {
+            if (queryRefFactory.getRelationId(column.getId()) == -1) {
+                return null;
+            }
+        }
+        ColumnEnforcer columnEnforcer = new ColumnEnforcer(query, columns);
+        OptExpression newQuery = columnEnforcer.enforce();
+        mvRewriteContext.setEnforcedColumns(columnEnforcer.getEnforcedColumns());
```

Comment:
what if enforce failed ?

Reply:
first, it has been checked in line 1420-1424.
secondly, if enforce failed, rewrite predicate will fail again, it will lead to rewrite failure.